From d28e816a99acb6b5453095ab9d89a746f092442d Mon Sep 17 00:00:00 2001
From: Tendai Makumire <121951760+tmakumire@users.noreply.github.com>
Date: Fri, 7 Feb 2025 13:12:50 +0000
Subject: [PATCH 1/2] Add support for booting IMDT SOM8550 in Linux (#7)

Add support for booting IMDT SBC8550 in Linux
To allow our board to boot without crashing
we had to remove the kernel drivers related to
the charger PMIC (PM8550B). We had support
from Qualcomm whilst making these changes.

I have also added a new feature to the repeater I2C
so that it requires no I2C Read/Writes. This was more
useful for another project where the eUSB2 repeater
was connected to I2C pins that weren't useable in Linux.
I can probably remove this change once I prove that APPS_I2C
is working with the eUSB repeater. For now I'll keep it here.
---
 arch/arm64/configs/vendor/kalama_GKI.config   | 14 +++++------
 .../configs/vendor/kalama_consolidate.config  |  2 +-
 .../vendor/kalama_le_consolidate.config       |  6 ++++-
 drivers/usb/repeater/repeater-i2c-eusb2.c     | 23 +++++++++++++++----
 4 files changed, 31 insertions(+), 14 deletions(-)

diff --git a/arch/arm64/configs/vendor/kalama_GKI.config b/arch/arm64/configs/vendor/kalama_GKI.config
index 453c6e8c9..c909160d3 100644
--- a/arch/arm64/configs/vendor/kalama_GKI.config
+++ b/arch/arm64/configs/vendor/kalama_GKI.config
@@ -64,7 +64,7 @@ CONFIG_I3C_MASTER_MSM_GENI=m
 CONFIG_IIO_BUFFER_CB=m
 CONFIG_INPUT_PM8941_PWRKEY=m
 # CONFIG_INPUT_PM8XXX_VIBRATOR is not set
-CONFIG_INPUT_QCOM_HV_HAPTICS=m
+# CONFIG_INPUT_QCOM_HV_HAPTICS is not set
 CONFIG_INTERCONNECT_QCOM=m
 CONFIG_INTERCONNECT_QCOM_BCM_VOTER=m
 CONFIG_INTERCONNECT_QCOM_CROW=m
@@ -277,12 +277,12 @@ CONFIG_QRTR_SMD=m
 CONFIG_QSEECOM_PROXY=m
 CONFIG_QSEE_IPC_IRQ_BRIDGE=m
 CONFIG_QTI_ALTMODE_GLINK=m
-CONFIG_QTI_BATTERY_CHARGER=m
-CONFIG_QTI_BATTERY_GLINK_DEBUG=m
+# CONFIG_QTI_BATTERY_CHARGER is not set
+# CONFIG_QTI_BATTERY_GLINK_DEBUG is not set 
 CONFIG_QTI_BCL_PMIC5=m
 CONFIG_QTI_BCL_SOC_DRIVER=m
 CONFIG_QTI_C1DCVS_SCMI_CLIENT=m
-CONFIG_QTI_CHARGER_ULOG_GLINK=m
+# CONFIG_QTI_CHARGER_ULOG_GLINK is not set 
 CONFIG_QTI_CPUFREQ_CDEV=m
 CONFIG_QTI_CPUFREQ_STATS_SCMI_CLIENT=m
 CONFIG_QTI_CPU_HOTPLUG_COOLING_DEVICE=m
@@ -296,10 +296,10 @@ CONFIG_QTI_HW_KEY_MANAGER=m
 CONFIG_QTI_HW_MEMLAT_SCMI_CLIENT=m
 CONFIG_QTI_IOMMU_SUPPORT=m
 CONFIG_QTI_PLH_SCMI_CLIENT=m
-CONFIG_QTI_PMIC_EUSB2_REPEATER=m
-CONFIG_QTI_PMIC_GLINK=m
+# CONFIG_QTI_PMIC_EUSB2_REPEATER is not set
+# CONFIG_QTI_PMIC_GLINK is not set
 # CONFIG_QTI_PMIC_GLINK_CLIENT_DEBUG is not set
-CONFIG_QTI_PMIC_GLINK_DEBUG=m
+# CONFIG_QTI_PMIC_GLINK_DEBUG is not set
 CONFIG_QTI_PMIC_PON_LOG=m
 CONFIG_QTI_PMU_SCMI_CLIENT=m
 CONFIG_QTI_POLICY_ENGINE_SENSOR=m
diff --git a/arch/arm64/configs/vendor/kalama_consolidate.config b/arch/arm64/configs/vendor/kalama_consolidate.config
index c5f2871e4..657fe3304 100644
--- a/arch/arm64/configs/vendor/kalama_consolidate.config
+++ b/arch/arm64/configs/vendor/kalama_consolidate.config
@@ -18,7 +18,7 @@ CONFIG_MSM_GPI_DMA_DEBUG=y
 # CONFIG_QCOM_DYN_MINIDUMP_STACK is not set
 CONFIG_QCOM_RTB=m
 CONFIG_QCOM_RTB_SEPARATE_CPUS=y
-CONFIG_QTI_PMIC_GLINK_CLIENT_DEBUG=y
+# CONFIG_QTI_PMIC_GLINK_CLIENT_DEBUG is not set
 CONFIG_RCU_TORTURE_TEST=m
 CONFIG_REGMAP_QTI_DEBUGFS_ALLOW_WRITE=y
 CONFIG_RPMSG_QCOM_GLINK_DEBUG=y
diff --git a/arch/arm64/configs/vendor/kalama_le_consolidate.config b/arch/arm64/configs/vendor/kalama_le_consolidate.config
index cdddb76fe..954a27548 100644
--- a/arch/arm64/configs/vendor/kalama_le_consolidate.config
+++ b/arch/arm64/configs/vendor/kalama_le_consolidate.config
@@ -17,7 +17,7 @@ CONFIG_POWER_SUPPLY_HWMON=y
 # CONFIG_QCOM_DYN_MINIDUMP_STACK is not set
 CONFIG_QCOM_RTB=m
 CONFIG_QCOM_RTB_SEPARATE_CPUS=y
-CONFIG_QTI_PMIC_GLINK_CLIENT_DEBUG=y
+# CONFIG_QTI_PMIC_GLINK_CLIENT_DEBUG is not set
 CONFIG_RCU_TORTURE_TEST=m
 CONFIG_REGMAP_QTI_DEBUGFS_ALLOW_WRITE=y
 CONFIG_RPMSG_QCOM_GLINK_DEBUG=y
@@ -28,3 +28,7 @@ CONFIG_SPI_SPIDEV=m
 CONFIG_TEST_USER_COPY=m
 CONFIG_THERMAL_HWMON=y
 CONFIG_UFS_DBG=y
+CONFIG_DEVMEM=y
+CONFIG_EXTCON=y
+CONFIG_USB_CONN_GPIO=m
+CONFIG_EXTCON_USB_GPIO=m
diff --git a/drivers/usb/repeater/repeater-i2c-eusb2.c b/drivers/usb/repeater/repeater-i2c-eusb2.c
index a20134a20..658721a76 100644
--- a/drivers/usb/repeater/repeater-i2c-eusb2.c
+++ b/drivers/usb/repeater/repeater-i2c-eusb2.c
@@ -70,6 +70,7 @@
 enum eusb2_repeater_type {
 	TI_REPEATER,
 	NXP_REPEATER,
+	IMDT_GPIO_REPEATER,
 };
 
 struct i2c_repeater_chip {
@@ -258,16 +259,21 @@ static int eusb2_repeater_init(struct usb_repeater *ur)
 	case NXP_REPEATER:
 		eusb2_i2c_read_reg(er, REVISION_ID, &reg_val);
 		break;
+	case IMDT_GPIO_REPEATER:
+		dev_info(er->ur.dev, "IMDT GPIO eUSB2 repeater!\n");
+		break;
 	default:
 		dev_err(er->ur.dev, "Invalid repeater\n");
 	}
 
-	dev_info(er->ur.dev, "eUSB2 repeater version = 0x%x ur->flags:0x%x\n", reg_val, ur->flags);
+	if (chip->repeater_type != IMDT_GPIO_REPEATER) {
+		dev_info(er->ur.dev, "eUSB2 repeater version = 0x%x ur->flags:0x%x\n", reg_val, ur->flags);
 
-	/* override init sequence using devicetree based values */
-	if (er->param_override_seq_cnt)
-		eusb2_repeater_update_seq(er, er->param_override_seq,
-					er->param_override_seq_cnt);
+		/* override init sequence using devicetree based values */
+		if (er->param_override_seq_cnt)
+			eusb2_repeater_update_seq(er, er->param_override_seq,
+						      er->param_override_seq_cnt);
+	}
 
 	dev_info(er->ur.dev, "eUSB2 repeater init\n");
 
@@ -307,6 +313,9 @@ static struct i2c_repeater_chip repeater_chip[] = {
 	},
 	[TI_REPEATER] = {
 		.repeater_type = TI_REPEATER,
+	},
+	[IMDT_GPIO_REPEATER] = {
+		.repeater_type = IMDT_GPIO_REPEATER,
 	}
 };
 
@@ -319,6 +328,10 @@ static const struct of_device_id eusb2_repeater_id_table[] = {
 		.compatible = "ti,eusb2-repeater",
 		.data = &repeater_chip[TI_REPEATER]
 	},
+	{
+		.compatible = "imdt,eusb2-repeater",
+		.data = &repeater_chip[IMDT_GPIO_REPEATER]
+	},
 	{ },
 };
 MODULE_DEVICE_TABLE(of, eusb2_repeater_id_table);
-- 
2.34.1

